### ISSUE TITLE
[MobileFaceNet]ONNX to tflite模型转换异常

### ISSUE THREAD

**Issue by joyoki** [MPolaris/onnx2tflite#18]
Hi 博主，
  感谢你提供的工具，我在转换mobilefacenet 模型的时候，遇到以下几个问题，不知能否帮忙解答：

**【问题描述】**
1.onnx2tflite之后，模型输出由128变成1000
```
//转换前：
name: fc1
type: float32[1,128]
```
```
//转换后：
name: StatefulPartitionedCall:0
type: float32[1,1000]
location: 177
```

2.将生成的模型导入APK验证，脸部检测功能能正常工作，但是但是人脸识别功能失效
此处暂未提供LOG，优先确认问题1

2.生成的tflite模型尺寸几乎和onnx大小相当，未见尺寸优化

```
(base) rsi@DESKTOP-1NGTB52:~/onnx2tflite/onnx2tflite/input$ ll
total 62388
drwxr-xr-x 2 rsi rsi     4096 Oct 31 11:39 ./
drwxr-xr-x 8 rsi rsi     4096 Oct 31 11:34 ../
-rw-r--r-- 1 rsi rsi 13965850 Oct 31 11:00 face_recognition_sface_2021dec-act_int8-wt_int8-quantized.onnx
-rw-r--r-- 1 rsi rsi 13979876 Oct 31 11:00 face_recognition_sface_2021dec-act_int8-wt_int8-quantized.tflite
-rw-r--r-- 1 rsi rsi 13965850 Oct 31 11:33 face_recognition_sface_2021dec.onnx
-rw-r--r-- 1 rsi rsi 13979876 Oct 31 11:34 face_recognition_sface_2021dec.tflite
```

2210311357更新：tflite模型偏大的问题不知道是不是该文档提到大的，onnx转tflite需要去冗余操作：https://zhuanlan.zhihu.com/p/363317178

【所使用转换脚本】
使用博主的参考py，改成如下onnx2tflite.py

```
import torch
import torchvision
_input = torch.randn(1, 3, 112, 112)
model = torchvision.models.mobilenet_v2(True)
# use default settings is ok
torch.onnx.export(model, _input, './input/face_recognition_sface_2021dec.onnx', opset_version=11)# or opset_version=13

from converter import onnx_converter
onnx_converter(
    onnx_model_path = "./input/face_recognition_sface_2021dec.onnx",
    need_simplify = True,
    output_path = "./",
    target_formats = ['tflite'], # or ['keras'], ['keras', 'tflite']
    weight_quant = False,
    int8_model = False,
    int8_mean = None,
    int8_std = None,
    image_root = None
)
```

**[相关模型]：**
1.我所转换的模型为[opencv_zoo](https://github.com/opencv/opencv_zoo)/[models](https://github.com/opencv/opencv_zoo/tree/master/models)/face_recognition_sface/
1.face_recognition_sface_2021dec-act_int8-wt_int8-quantized.onnx
2.[face_recognition_sface_2021dec.onnx](https://github.com/opencv/opencv_zoo/blob/master/models/face_recognition_sface/face_recognition_sface_2021dec.onnx)
原始模型链接为：https://github.com/opencv/opencv_zoo/tree/master/models/face_recognition_sface

谢谢解答

**Comment by MPolaris** [comment]
输出128变1000是因为你导出的是mobilenet, 
```python 
# 你导出的onnx就是mobilenet
model = torchvision.models.mobilenet_v2(True)
```

**Comment by joyoki** [comment]
> 输出128变1000是因为你导出的是mobilenet,
> 
> ```python
> # 你导出的onnx就是mobilenet
> model = torchvision.models.mobilenet_v2(True)
> ```

感谢博主的指正，请问这块应该改成什么样呢？另外是否还有其他错误，也辛苦博主帮忙指正？

**Comment by MPolaris** [comment]
> > 输出128变1000是因为你导出的是mobilenet,
> > ```python
> > # 你导出的onnx就是mobilenet
> > model = torchvision.models.mobilenet_v2(True)
> > ```
> 
> 感谢博主的指正，请问这块应该改成什么样呢？有没有对应的model输出要求的必改格式对照表，或者置为缺省，是否可以根据导入的模型，自动转换？
> 
> 另外是否还有其他错误，也辛苦博主帮忙指正？
先重新pull一下代码，然后
```python
from converter import onnx_converter
onnx_converter(
    onnx_model_path = "./input/face_recognition_sface_2021dec.onnx",#这里填你的模型地址
    need_simplify = True,
    output_path = "./",
    target_formats = ['tflite'], # or ['keras'], ['keras', 'tflite']
    weight_quant = False,
    int8_model = False,
    int8_mean = None,
    int8_std = None,
    image_root = None
)
```

**Comment by joyoki** [comment]
> 

感谢解答，我将代码同步到最新，尝试博主提供的方法，还是遇到相同的output=1000的问题，我在另外一个github上看到一个onnx2tf的工具也是引用了博主的套件，感觉博主的工具很有潜力，不知博主是否方便继续帮忙确认下？问题具体描述如下：
1.我当前的代码节点已同步到最新：
```
//当前代码节点：
(base) rsi@DESKTOP-1NGTB52:~/onnx2tflite/onnx2tflite$ git log
commit 24b6647c97ca0a74fb8965e0929e4e0bf6775bb4 (HEAD, origin/main, origin/HEAD)
Author: MPolaris <540492239@qq.com>
Date:   Mon Oct 31 13:54:51 2022 +0800

    update parse arguments

commit 8e7451def93c3f1886e7cfd3b54678856d93f94b
Author: MPolaris <540492239@qq.com>
Date:   Mon Oct 31 13:53:49 2022 +0800
...
```
转换所用脚本onnx2tflite.py
```
//onnx2tflite转换脚本代码:
# git clone https://github.com/MPolaris/onnx2tflite
# cd onnx2tflite
# pip install -r requirements.txt
 
#import torch
#import torchvision
#_input = torch.randn(1, 3, 112, 112)
#model = torchvision.models.mobilenet_v2(True)
## use default settings is ok
#torch.onnx.export(model, _input, './input/face_recognition_sface_2021dec.onnx', opset_version=11)# or opset_version=13
 
from converter import onnx_converter
onnx_converter(
    onnx_model_path = "./input/face_recognition_sface_2021dec.onnx",#这里填你的模型地址
    need_simplify = True,
    output_path = "./input",
    target_formats = ['tflite'], # or ['keras'], ['keras', 'tflite']
    weight_quant = False,
    int8_model = False,
    int8_mean = None,
    int8_std = None,
    image_root = None
)
```
//转换结果异常部分：
```
StatefulPartitionedCall:0177
name: StatefulPartitionedCall:0
type: float32[1,1000]  //此处应该为[1,128]
location: 177
```

另外我在另外一个github上看到一个组合工具：https://github.com/PINTO0309/onnx2tf  看到也是引用博主的工具，但是使用该工具，可以将onnx转换为tflite（但是从测试结果看，转换后精度不高，这个问题稍后再表）

```
//转换命令：
onnx2tf -i face_recognition_sface_2021dec.onnx -o saved_model
```
```
//转换后的outputs:
Identity295
name: Identity
type: float32[1,128]
location: 295
```

谢谢博主

**Comment by MPolaris** [comment]
> > 
> 
> 感谢解答，我将代码同步到最新，尝试博主提供的方法，还是遇到相同的output=1000的问题，我在另外一个github上看到一个onnx2tf的工具也是引用了博主的套件，感觉博主的工具很有潜力，不知博主是否方便继续帮忙确认下？问题具体描述如下： 1.我当前的代码节点已同步到最新：
> 
> ```
> //当前代码节点：
> (base) rsi@DESKTOP-1NGTB52:~/onnx2tflite/onnx2tflite$ git log
> commit 24b6647c97ca0a74fb8965e0929e4e0bf6775bb4 (HEAD, origin/main, origin/HEAD)
> Author: MPolaris <540492239@qq.com>
> Date:   Mon Oct 31 13:54:51 2022 +0800
> 
>     update parse arguments
> 
> commit 8e7451def93c3f1886e7cfd3b54678856d93f94b
> Author: MPolaris <540492239@qq.com>
> Date:   Mon Oct 31 13:53:49 2022 +0800
> ...
> ```
> 
> 转换所用脚本onnx2tflite.py
> 
> ```
> //onnx2tflite转换脚本代码:
> # git clone https://github.com/MPolaris/onnx2tflite
> # cd onnx2tflite
> # pip install -r requirements.txt
>  
> #import torch
> #import torchvision
> #_input = torch.randn(1, 3, 112, 112)
> #model = torchvision.models.mobilenet_v2(True)
> ## use default settings is ok
> #torch.onnx.export(model, _input, './input/face_recognition_sface_2021dec.onnx', opset_version=11)# or opset_version=13
>  
> from converter import onnx_converter
> onnx_converter(
>     onnx_model_path = "./input/face_recognition_sface_2021dec.onnx",#这里填你的模型地址
>     need_simplify = True,
>     output_path = "./input",
>     target_formats = ['tflite'], # or ['keras'], ['keras', 'tflite']
>     weight_quant = False,
>     int8_model = False,
>     int8_mean = None,
>     int8_std = None,
>     image_root = None
> )
> ```
> 
> //转换结果异常部分：
> 
> ```
> StatefulPartitionedCall:0177
> name: StatefulPartitionedCall:0
> type: float32[1,1000]  //此处应该为[1,128]
> location: 177
> ```
> 
> 另外我在另外一个github上看到一个组合工具：https://github.com/PINTO0309/onnx2tf 看到也是引用博主的工具，但是使用该工具，可以将onnx转换为tflite（但是从测试结果看，转换后精度不高，这个问题稍后再表）
> 
> ```
> //转换命令：
> onnx2tf -i face_recognition_sface_2021dec.onnx -o saved_model
> ```
> 
> ```
> //转换后的outputs:
> Identity295
> name: Identity
> type: float32[1,128]
> location: 295
> ```
> 
> 谢谢博主

朋友，你还是转换的mobilenet，只是你把模型的名字写成了face_recognition_sface_2021dec.onnx。
你要不要试试把./input/face_recognition_sface_2021dec.onnx删掉，然后重新下载一个？

**Comment by joyoki** [comment]
@MPolaris 
博主，实在抱歉，耽误你时间了。
对于这个参考你wiki example写的转换脚本onnx2tflite.py还是不太理解。能否麻烦博主用这个模型face_recognition_sface_2021dec.onnx提供个完整的参考？十分感谢

```
//我已经将你提供的参考里面的mobilenet 注释掉
朋友，你还是转换的mobilenet，只是你把模型的名字写成了face_recognition_sface_2021dec.onnx。
你要不要试试把./input/face_recognition_sface_2021dec.onnx删掉，然后重新下载一个？
```

转换所用脚本onnx2tflite.py：

```
//onnx2tflite转换脚本代码:
# git clone https://github.com/MPolaris/onnx2tflite
# cd onnx2tflite
# pip install -r requirements.txt
 
#import torch
#import torchvision
#_input = torch.randn(1, 3, 112, 112)
#model = torchvision.models.mobilenet_v2(True)
## use default settings is ok
#torch.onnx.export(model, _input, './input/face_recognition_sface_2021dec.onnx', opset_version=11)# or opset_version=13
 
from converter import onnx_converter
onnx_converter(
    onnx_model_path = "./input/face_recognition_sface_2021dec.onnx",#这里填你的模型地址
    need_simplify = True,
    output_path = "./input",
    target_formats = ['tflite'], # or ['keras'], ['keras', 'tflite']
    weight_quant = False,
    int8_model = False,
    int8_mean = None,
    int8_std = None,
    image_root = None
)
```
谢谢

**Comment by PINTO0309** [comment]
@joyoki
It appears to be a closed issue, so you can ignore it. I do not intend to disturb you all.
The onnx2tf already supports transforms for almost all OPs except RNNs, and also supports INT8 quantization, INT8 with INT16 activation and Dynamic Range Quantization.
https://github.com/PINTO0309/onnx2tf

The reason why @joyoki said the accuracy is poor after conversion is clear. You may have been missing a parameter that should have been specified to onnx2tf. I have included the URL of the tflite model I converted so you can compare it to the model you converted.

- `face_recognition_sface_2021dec.tflite` float32/float16/int8/int8 with int16 act ...
https://drive.google.com/file/d/1_XPuPQNcWTj3vx3yb0mPuA2O63HdEGul/view?usp=share_link

- replace.json https://github.com/PINTO0309/onnx2tf#parameter-replacement
  ```json
  {
    "format_version": 1,
    "operations": [
      {
        "op_name": "flatten_254",
        "param_target": "inputs",
        "param_name": "dropout5",
        "pre_process_transpose_perm": [0,3,1,2]
      }
    ]
  }
  ```
- convert command
  ```bash
  $ onnx2tf -i face_recognition_sface_2021dec.onnx -prf replace.json
  ```

- model
  face_recognition_sface_2021dec.onnx

- ONNX Flatten
![image](https://user-images.githubusercontent.com/33194443/199764265-2318f8e6-fab4-46bf-a563-9ab8c9f30639.png)

- My TFLite Flatten
![image](https://user-images.githubusercontent.com/33194443/199764927-727f0477-0444-49c5-aa17-a25c284f15cd.png)

Thank you very much for your time.
